app/controllers/Users.scala

```diff
@@ -76,17 +77,16 @@ class Users @Inject()(fakeUser: FakeUser,
       Future.successful(redirectToSso(this.sso.getLoginUrl(this.baseUrl + "/login", nonce)))
     } else {
       // Redirected from SpongeSSO, decode SSO payload and convert to Ore user
-      this.sso.authenticate(sso.get, sig.get)(isNonceValid) flatMap {
-        case None =>
-          Future.successful(Redirect(ShowHome).withError("error.loginFailed"))
-        case Some(spongeUser) =>
-          // Complete authentication
-          User.fromSponge(spongeUser).flatMap(this.users.getOrCreate).flatMap { user =>
-            user.pullForumData().flatMap(_.pullSpongeData()).flatMap { u =>
-              this.redirectBack(request.flash.get("url").getOrElse("/"), u)
-            }
-          }
-      }
+      this.sso.authenticate(sso.get, sig.get)(isNonceValid).semiFlatMap { spongeUser =>
+        // Complete authentication
+        for {
+          fromSponge <- User.fromSponge(spongeUser)
+          getOrCreate <- this.users.getOrCreate(fromSponge)
+          pulledForum <- getOrCreate.pullForumData()   //TODO These two pull methods at the moment just return this at the end.
```

Comment:
This doesn't have to be returned. It updates the user data inside the pull function. We can ignore their result IMO.